### PR TITLE
Space - Hala 

### PR DESCRIPTION
**Pull request by halahaddad1**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The model manages the database 
Describe in your own words what the Controller is doing in Rails | It coordinates the interactions between the view and the models
Describe in your own words what the View is doing in Rails | Shares data with the browser and the controller
Describe an edge-case controller test you wrote | One example is to test that the test saves a new task in the create task page, and what happens if it doesn't save.
What is the purpose of using strong params? (i.e. the params method in the controller) | It's the rails way of getting saved variables to our controller actions.
How are Rails migrations related to Rails models? | Migrations update our database, the database updates rely on the model to specify the layout, so migrations are streamlined by the model.
Describe one area of Rails that are still unclear on | I think everything in rails is still very much raw and writing code feels like stepping on egg shells, every step breaks something, it will take more practice to be sure about more complex actions. the toggle complete/not complete action for example took a whole day for me to figure out! looking forward to becoming more confidant writing actions in rails.

### CODE REVIEW

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# TASKS = [
+#     {task: "Clean the dishes"},
+#     {task: "Take out the trash"},
+#     {task: "Take the kids to the park"}
+# ]
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:5]:
```suggestion
```

Can be deleted

File: app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        @task = Task.new
+    end
+
+    def create
+        @task = Task.new(name: params[:task][:name],description: params[:task][:description],completed_at: params[:task][:completed_at]) 
+        if @task.save
+          redirect_to task_path(@task.id)
+          return
+        else 
+          render :new
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:22]: Just a suggestion, include a status code:

```suggestion
          render :new, status: :bad_request
```

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            return 
+        elsif @task.update(
+            name: params[:task][:name],
+            description: params[:task][:description],
+            completed_at: params[:task][:completed_at]
+        )
+            redirect_to tasks_path
+            return
+        else
+            render :edit
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:57]:
```suggestion
            render :edit, status: :bad_request
```

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        end
+    end
+
+    def update_completed_task
+        @task= Task.find_by(id: params[:id])
+        if @task.completed_at == nil
+            if @task.update(
+                completed_at: Date.today.to_s
+            ) 
+                redirect_to request.referrer
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:68]: Nice!

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  <head>
+    <title>TaskList</title>
+    <%= csrf_meta_tags %>
+    <%= csp_meta_tag %>
+
+    <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+  </head>
+
+  <body>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:12]: I suggest putting navigation links here like maybe a link to create a new task, and view all tasks.

File: app/views/tasks/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<%= form_with model: @task, class: 'edit-task' do |f| %>
```
Comment (CheezItMan) [inline app/views/tasks/edit.html.erb:1]: I know this isn't required for Task List, just noting that for future projects you can DRY up this form code with a partial view.

File: app/views/tasks/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<h1 class="bg-info text-white text">Current Tasks:<h1>
+        <div class="">
+            <% @tasks.each do |task| %>
+            <div class="dl-horizontal">
+                <% css_class = (task.completed_at.nil? || task.completed_at.empty?) ? 'nostrike' : 'strike' %>
```
Comment (CheezItMan) [inline app/views/tasks/index.html.erb:5]: I like this.

File: app/views/tasks/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<div>
+    <h1><%= @task.name %></h1>
+    <p>: <%= @task.description %></p>
+    <p> <%= @task.completed_at %></p>
+    <%= link_to "Edit", edit_task_path(@task), class:"btn btn-info" %>
```
Comment (CheezItMan) [inline app/views/tasks/show.html.erb:5]: I suggest using the strikethrough styling here as well.

File: app/controllers/tasks_controller.rb [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        )
+            redirect_to tasks_path
+            return
+        else
+            render :edit
+            return
+        end
+    end
+
+    def update_completed_task
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:62]: I also like the toggle.

**Comment by CheezItMan** [comment]
# Task List




## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| At least 6 commits with meaningful commit messages | ✔️
| Routes follow RESTful conventions | ✔️
| Uses named routes (like `_path`) | ✔️
| Creates Models and migrations | ✔️
| Creates styled views | ✔️
| Handles errors like nonexistant tasks | ✔️
| Uses `form_with` to render forms in Rails | ✔️

## Functional Requirements/Manual Testing

| Functional Requirement | yes/no |
| --- | --- |
| Successfully handles index & show | ✔️
| index & show tests pass | ✔️
| Successfully handles: New, Create | ✔️
| New, Create tests pass | ✔️
| Successfully handles: Edit, Update | ✔️
| Successfully handles: Destroy, Task Complete | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 5+ in Functional Requirements |






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅



## Summary


Nicely done, you hit all the learning goals here.  I like the styling you added here.  Well done.
